internal/app/machined/pkg/controllers/hardware/pci_driver_rebind.go

```diff
@@ -0,0 +1,180 @@
+	v1alpha1runtime "github.com/siderolabs/talos/internal/app/machined/pkg/runtime"
+	"github.com/siderolabs/talos/pkg/machinery/resources/hardware"
+)
+
+const (
+	targetDeviceSYSFSPath = "/sys/bus/pci/devices/%s"
+	driverOverridePath    = targetDeviceSYSFSPath + "/driver_override"
+	driverUnbindPath      = targetDeviceSYSFSPath + "/driver/unbind"
+	driverPath            = targetDeviceSYSFSPath + "/driver"
+	driverProbePath       = "/sys/bus/pci/drivers_probe"
+)
+
+// PCIDriverRebindController binds PCI devices to a specific driver and unbinds them from the host driver.
+type PCIDriverRebindController struct {
+	V1Alpha1Mode v1alpha1runtime.Mode
+
+	boundDevices map[string]struct{}
+}
+
+// Name implements controller.Controller interface.
+func (c *PCIDriverRebindController) Name() string {
+	return "hardware.PCIDriverRebindController"
+}
+
+// Inputs implements controller.Controller interface.
+func (c *PCIDriverRebindController) Inputs() []controller.Input {
+	return nil
+}
+
+// Outputs implements controller.Controller interface.
+func (c *PCIDriverRebindController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: hardware.PCIDriverRebindStatusType,
+			Kind: controller.OutputExclusive,
+		},
+	}
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo
+func (c *PCIDriverRebindController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) (err error) {
+	// Skip PCI rebind handling if running in a container or agent mode.
+	if c.V1Alpha1Mode.InContainer() || c.V1Alpha1Mode.IsAgent() {
+		return nil
+	}
+
+	if c.boundDevices == nil {
+		c.boundDevices = map[string]struct{}{}
+	}
+
+	// wait for udevd to be healthy, this is to ensure that host drivers if any are loaded.
+	if err := runtimectrl.WaitForDevicesReady(ctx, r,
+		[]controller.Input{
+			{
+				Namespace: hardware.NamespaceName,
+				Type:      hardware.PCIDriverRebindConfigType,
+				Kind:      controller.InputWeak,
+			},
+		}); err != nil {
+		return fmt.Errorf("error waiting for devices to be ready: %w", err)
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+		}
+
+		pciDriverRebindConfigs, err := safe.ReaderListAll[*hardware.PCIDriverRebindConfig](ctx, r)
+		if err != nil {
+			return fmt.Errorf("error listing all PCI rebind configs: %w", err)
+		}
+
+		r.StartTrackingOutputs()
+
+		touchedIDs := map[string]struct{}{}
+
+		for cfg := range pciDriverRebindConfigs.All() {
+			if err := c.handlePCIDriverReBind(cfg.TypedSpec().PCIID, cfg.TypedSpec().TargetDriver); err != nil {
+				return err
+			}
+
+			boundDriver, err := checkDeviceBoundDriver(cfg.TypedSpec().PCIID)
+			if err != nil {
+				return fmt.Errorf("error checking bound driver for device with id: %s, %w", cfg.TypedSpec().PCIID, err)
+			}
+
+			if boundDriver != cfg.TypedSpec().TargetDriver {
+				logger.Info(
+					"cannot validate if device is bound to target driver, ensure target driver module is loaded",
```

Comment:
on second thoughts i think we should error, otherwise if the module is loaded later, we cannot guarantee that the driver is rebinded, also this would skip creating the status resource

Reply:
actually this is fine, when module is loaded later on, the driver gets bind